File: Platforms/Common/Reign_Video_D3D11_Component/States/RasterizerStateDesc.cpp
#include "pch.h"
#include "RasterizerState.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	RasterizerStateDescCom::RasterizerStateDescCom(REIGN_D3D11_FILL_MODE fillMode, REIGN_D3D11_CULL_NONE cullMode, bool frontCounterClockwise, int depthBias, float depthBiasClamp, float slopeScaledDepthBias, bool depthClipEnable, bool scissorEnable, bool multisampleEnable, bool antialiasedLineEnable)
	{
		null();

		desc = new D3D11_RASTERIZER_DESC();
		ZeroMemory(desc, sizeof(D3D11_RASTERIZER_DESC));

		desc->FillMode = (D3D11_FILL_MODE)fillMode;
		desc->CullMode = (D3D11_CULL_MODE)cullMode;
		desc->FrontCounterClockwise = frontCounterClockwise;
		desc->DepthBias = depthBias;
		desc->DepthBiasClamp = depthBiasClamp;
		desc->SlopeScaledDepthBias = slopeScaledDepthBias;
		desc->DepthClipEnable = depthClipEnable;
		desc->ScissorEnable = scissorEnable;
		desc->MultisampleEnable = multisampleEnable;
		desc->AntialiasedLineEnable = antialiasedLineEnable;
	}
	#pragma endregion

	#pragma region Methods
	RasterizerStateDescCom::~RasterizerStateDescCom()
	{
		if (desc) delete desc;
		null();
	}

	void RasterizerStateDescCom::null()
	{
		desc = 0;
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/States/RasterizerStateDesc.cs
using System;
using Reign.Core;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class RasterizerStateDesc : RasterizerStateDescI
	{
		#region Properties
		internal RasterizerStateDescCom com;
		#endregion

		#region Constructors
		public static RasterizerStateDesc New(RasterizerStateTypes type)
		{
			return new RasterizerStateDesc(type);
		}

		public RasterizerStateDesc(RasterizerStateTypes type)
		{
			try
			{
				REIGN_D3D11_FILL_MODE fillMode = REIGN_D3D11_FILL_MODE.SOLID;
				REIGN_D3D11_CULL_NONE cullMode = REIGN_D3D11_CULL_NONE.NONE;
				bool frontCounterClockwise = false, depthClipEnable = false, scissorEnable = false, multisampleEnable = false, antialiasedLineEnable = false;
				int depthBias = 0;
				float depthBiasClamp = 0, slopeScaledDepthBias = 0;
				switch (type)
				{
					case RasterizerStateTypes.Solid_CullNone:
						fillMode = REIGN_D3D11_FILL_MODE.SOLID;
						cullMode = REIGN_D3D11_CULL_NONE.NONE;
						frontCounterClockwise = false;
						depthBias = 0;
						depthBiasClamp = 0.0f;
						slopeScaledDepthBias = 0.0f;
						depthClipEnable = true;
						scissorEnable = false;
						multisampleEnable = false;
						antialiasedLineEnable = false;
						break;

					case RasterizerStateTypes.Solid_CullCW:
						fillMode = REIGN_D3D11_FILL_MODE.SOLID;
						cullMode = REIGN_D3D11_CULL_NONE.FRONT;
						frontCounterClockwise = false;
						depthBias = 0;
						depthBiasClamp = 0.0f;
						slopeScaledDepthBias = 0.0f;
						depthClipEnable = true;
						scissorEnable = false;
						multisampleEnable = false;
						antialiasedLineEnable = false;
						break;

					case RasterizerStateTypes.Solid_CullCCW:
						fillMode = REIGN_D3D11_FILL_MODE.SOLID;
						cullMode = REIGN_D3D11_CULL_NONE.BACK;
						frontCounterClockwise = false;
						depthBias = 0;
						depthBiasClamp = 0.0f;
						slopeScaledDepthBias = 0.0f;
						depthClipEnable = true;
						scissorEnable = false;
						multisampleEnable = false;
						antialiasedLineEnable = false;
						break;

					default:
						Debug.ThrowError("RasterizerStateDesc", "Unsuported type: " + type.ToString());
						break;
				}

				com = new RasterizerStateDescCom(fillMode, cullMode, frontCounterClockwise, depthBias, depthBiasClamp, slopeScaledDepthBias, depthClipEnable, scissorEnable, multisampleEnable, antialiasedLineEnable);
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public void Dispose()
		{
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
		}
		#endregion
	}
}